Re-writing accessors transformer using the new AGP API

nhachicha opened the pull request:
- [x] Migrating the Realm accessor Transformer to use the new AHP transformer API
- [x] Migrated Gradle plugin from Groovy to Kotlin.
- [x] Re-writing the internal object server transformer

Closes #7714

iamlordalvarado commented on dependencies.list:
```diff
@@ -7,10 +7,10 @@ REALM_CORE=11.14.0
 ANDROID_BUILD_TOOLS=30.0.3
-KOTLIN=1.5.31
-KOTLIN_COROUTINES=1.5.2
+KOTLIN=1.6.20
```
Probably you can try 1.7 as is the latest stable available

iamlordalvarado commented on examples/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.3.3-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.4-bin.zip
```
7.5 is also available

nhachicha commented:
@rorbech I updated the current WIP
- Currently, the JNI generated header files (.h) are not created if we invoke `assemble`, this might be caused by the various bumps to Gradle and AGP ... if you run `compileBaseReleaseJavaWithJavac` for instance then it will compile and generate the headers. 
- The internal transformer (the one that strips the ObjectServer symbols) needs to be re-written similarly to the accessor one.
- I updated some tests projects to run the new transformer against them
- We need to create a feature branch and release from it (probably using this template https://github.com/realm/realm-java/pull/7704

For reference:
- https://issuetracker.google.com/issues/237190740#comment4

rorbech commented:
I manually verified that all our `examples` are running with the current state. Also verified that updating model and other sources is properly rebuilt in incremental builds for a couple of the examples.

cmelchior commented on library-build-transformer/src/main/kotlin/io/realm/buildtransformer/RealmBuildTransformer.kt:
```diff
@@ -39,123 +52,124 @@ val logger: Logger = LoggerFactory.getLogger("realm-build-logger")
+    private val output: RegularFileProperty,
+) {
+    init {
+        transform()
```
Does all work really happen in the constructor? That feels a bit surprising IMO. It would maybe be nicer if there was some kind of `run` method, so we split constructing and running the class.

rorbech replied:
I moved just exposed `transform` as `internal` and call it on the constructed object instead of implicitly from the `init`-block.

cmelchior commented on library-build-transformer/src/main/kotlin/io/realm/buildtransformer/RealmBuildTransformer.kt:
```diff
@@ -39,123 +52,124 @@ val logger: Logger = LoggerFactory.getLogger("realm-build-logger")
+
+        forEachJarEntry { jarEntry, inputStream ->
+            if (jarEntry.name.endsWith(".class") ) {
+            inputStream.use {
```
Indentation

cmelchior commented on library-build-transformer/src/main/kotlin/io/realm/buildtransformer/RealmBuildTransformer.kt:
```diff
@@ -39,123 +52,124 @@ val logger: Logger = LoggerFactory.getLogger("realm-build-logger")
+        val metadataCollector =
+            io.realm.buildtransformer.asm.visitors.AnnotationVisitor(annotationDescriptor)
+
+        forEachJarEntry { jarEntry, inputStream ->
```
It is a bit unclear why we iterate the jar entries twice, some documentation for each block would be nice.

rorbech replied:
Added some comments

cmelchior commented on library-build-transformer/src/main/kotlin/io/realm/buildtransformer/RealmBuildTransformer.kt:
```diff
@@ -39,123 +52,124 @@ val logger: Logger = LoggerFactory.getLogger("realm-build-logger")
-                    return@iterateDirs // Directory was deleted
+        forEachJarEntry { jarEntry, inputStream ->
+            val bytes = inputStream.use { inputStream ->
+                if (jarEntry.name.endsWith(".class")) {
```
What happens with ressource files in the JAR? We seem to skip them here, but I'm not sure how they then end up in the outputprovider?

rorbech replied:
They are handled by the else branch 🎉 🤪

cmelchior commented on realm-transformer/src/main/kotlin/io/realm/transformer/RealmTransformer.kt:
```diff
@@ -43,56 +55,64 @@ data class ProjectMetaData(
     private lateinit var metadata: ProjectMetaData
     private var analytics: RealmAnalytics? = null
 
     init {
```
Same, running all of this logic directly in the constructor feels ...iffy

rorbech replied:
Same, I moved just exposed transform as internal and call it on the constructed object instead of implicitly from the init-block.

cmelchior commented on realm-transformer/src/main/kotlin/io/realm/transformer/RealmTransformer.kt:
```diff
@@ -43,56 +55,64 @@ data class ProjectMetaData(
+
+        fun register(project: Project) {
+            val androidComponents = project.extensions.getByType(AndroidComponentsExtension::class.java)
+            androidComponents.onVariants { variant ->
```
Magic 🙈, but if it works...

cmelchior commented on realm-transformer/src/main/kotlin/io/realm/transformer/build/BuildTemplate.kt:
```diff
@@ -107,51 +111,46 @@ abstract class BuildTemplate(val metadata: ProjectMetaData, val outputProvider:
-        classPool.close();
+    fun copyProcessedClasses() {
+        for ((fqname: String, clazz: CtClass) in processedClasses) {
+            outputProvider.putNextEntry(JarEntry("${fqname.replace('.', '/')}.class"))
```
Does this also work on Windows or should we use File.separatorChar?

rorbech replied:
I would argue heavily that the Zip file entry header should be platform agnostic ... otherwise you would need to supply different Zip-files for each platform. And since it works on one platform then I assume that it would also work on the others.

cmelchior reviewed (approved):
Awesome work 🥇 🕺  ... As I don't really know the new transform API it was a bit hard to track some of the changes, but if all our sample projects are building and tests are running, I think we can assume that things are working 🥳 

I only left some minor comments.

cmelchior reviewed (approved):
💯
